rfd/0153-resource-guidelines.md

```diff
@@ -0,0 +1,610 @@
+them, see [#29234](https://github.com/gravitational/teleport/issues/29234). In addition to reducing confusion and
+fostering a better developer experience, this will also remove complexity from the terraform provider and teleport
+operator that have to accommodate for every subtle API and resource difference that exist today.
+
+## Details
+
+### Defining a resource
+
+A resource at minimum MUST include a kind, version, `teleport.header.v1.Metadata` and a specification message with any
+additional parameters needed to represent the resource. While the kind and version may seem like they would be easy to
+derive from the message definition itself, they need to be defined so that anything processing a generic resource can
+identify which resource is being processed. For example, `tctl` interacts with resources in their in raw yaml text form
+and leverage `services.UnknownResource` to identify the resource and act appropriately.
+
+```protobuf
+import "teleport/header/v1/resourceheader.proto";
+
+// Foo is a resource that does foo.
+message Foo {
+  // The kind of resource represented.
+  string kind = 1;
+  // An optional subkind to differentiate variations of the same kind.
+  string sub_kind = 2;
+  // The version of the resource being represented.
+  string version = 3;
+  // Common metadata that all resources shared.
+  teleport.header.v1.Metadata metadata = 4;
+  // The resource specific specification.
+  FooSpec spec = 5;
+}
+
+// FooSpec contains resource specific properties.
+message FooSpec {
+  string bar = 1;
+  int32 baz = 2;
+  bool qux = 3;
+}
+```
+
+This differs from existing resources because legacy resources make heavy use of features provided
+by [gogoprotobuf](https://github.com/gogo/protobuf). Since that project has long been abandoned, we're striving to
+migrate away from it as described in [RFD-0139](https://github.com/gravitational/teleport/pull/28386).
+The `teleport.header.v1.Metadata` is a clone of `types.Metadata` which doesn't use any of the gogoproto features.
+Legacy resources also had a `types.ResourceHeader` that used gogo magic to embed the type in the resource message. To
+get around this, the required fields from the header MUST be included in the message itself. A non-gogo clone does exist
+`teleport.header.v1.ResourceHeader`, however, to get the fields, embeded custom marshalling must be manually written.
+
+If a resource has associated secrets (password, private key, jwt, mfa device, etc.) they should be defined in a separate
+resource and stored in a separate key range in the backend. The traditional pattern of defining secrets inline and only
+returning them if a `with_sercrets` flag is provided cause a variety of problems and introduce opportunity for human
+error to accidentally include secrets when they should not have been. It would then be the responsibility of the caller
+to get both the base resource and the corresponding secret resource if required.
+
```

Comment:
Bump? I think proto->json conversion and whether we use protos as-is for tctl and the like are still somewhat open.

For example, device trust defines "presentation" protos for tctl, but I'm not sure I would do that again today.

Reply:
> If a resource includes an enum but the desire is for tctl to display a friendly string representation instead of a number then I'd advise to use a presentation type

That was my main motivation, but today I would even question that. The reality is that we have plenty of types that expose numerical enums already, so the work felt like a drop in the ocean (but nevertheless took time and effort).

WDYT? Is this an area we should delve into? (Maybe not now but in a future update of the RFD?)